// @intility/bifrost-react/dist/components/Navigation/Nav/Nav.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _Nav = _interopRequireDefault(require("./Nav.Top"));

var _Nav2 = _interopRequireDefault(require("./Nav.Side"));

var _Nav3 = _interopRequireDefault(require("./Nav.Mobile"));

var _Nav4 = _interopRequireDefault(require("./Nav.Item"));

var _Nav5 = _interopRequireDefault(require("./Nav.Group"));

var _Nav6 = _interopRequireDefault(require("./Nav.Header"));

var _Nav7 = _interopRequireDefault(require("./Nav.Search"));

var _NavContext = _interopRequireDefault(require("./NavContext.internal"));

var _classnames = _interopRequireDefault(require("classnames"));

var _useBreakpoint = _interopRequireDefault(require("../../Utility/Breakpoint/useBreakpoint"));

var _ = require("../../..");

var _Nav8 = _interopRequireDefault(require("./Nav.Theme"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Responsive component for Bifrost navigation menus
 */
var Nav = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _sideProps$collapsed, _sideProps$onCollapse;

  var children = _ref.children,
      appName = _ref.appName,
      top = _ref.top,
      topProps = _ref.topProps,
      side = _ref.side,
      sideProps = _ref.sideProps,
      mobile = _ref.mobile,
      mobileProps = _ref.mobileProps,
      hideCollapse = _ref.hideCollapse,
      collapsedAppName = _ref.collapsedAppName,
      hideTheme = _ref.hideTheme,
      logoHref = _ref.logoHref,
      logoOnClick = _ref.logoOnClick,
      props = _objectWithoutProperties(_ref, ["children", "appName", "top", "topProps", "side", "sideProps", "mobile", "mobileProps", "hideCollapse", "collapsedAppName", "hideTheme", "logoHref", "logoOnClick"]);

  (0, _.useTheme)();

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      mobileOpen = _useState2[0],
      setMobileOpen = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      hasToggledNavMobile = _useState4[0],
      setHasToggledNavMobile = _useState4[1];

  var isMobile = (0, _useBreakpoint.default)(null, 'xl');
  (0, _react.useEffect)(function () {
    if (typeof document === 'undefined') return;

    function closeMobileNavOnEsc(e) {
      if (e.key === 'Escape') setMobileOpen(false);
    }

    document.addEventListener('keydown', closeMobileNavOnEsc);
    return function () {
      return document.removeEventListener('keydown', closeMobileNavOnEsc);
    };
  }, [setMobileOpen]);

  var _useLocalStorage = (0, _.useLocalStorage)('bfNavCollapsed', false),
      _useLocalStorage2 = _slicedToArray(_useLocalStorage, 2),
      collapsedState = _useLocalStorage2[0],
      setCollapsedState = _useLocalStorage2[1];

  var collapsed = (_sideProps$collapsed = sideProps === null || sideProps === void 0 ? void 0 : sideProps.collapsed) !== null && _sideProps$collapsed !== void 0 ? _sideProps$collapsed : collapsedState;
  var setCollapsed = (_sideProps$onCollapse = sideProps === null || sideProps === void 0 ? void 0 : sideProps.onCollapsedChange) !== null && _sideProps$onCollapse !== void 0 ? _sideProps$onCollapse : setCollapsedState; // Hiding body scrollbar when nav mobile is open

  (0, _react.useEffect)(function () {
    if (typeof window === 'undefined') return;

    if (isMobile && mobileOpen && hasToggledNavMobile) {
      // document.body.style.overflowY = 'hidden'
      document.body.classList.add('bf-nav-mobile-open');
    } else {
      // document.body.style.overflowY = 'auto'
      document.body.classList.remove('bf-nav-mobile-open');
    }
  }, [isMobile, mobileOpen]); // only apply animations after the first time mobileOpen is set to true

  (0, _react.useEffect)(function () {
    if (!mobileOpen) return;
    setHasToggledNavMobile(true);
  }, [mobileOpen]);
  return /*#__PURE__*/_react.default.createElement(_NavContext.default, {
    sideCollapsed: collapsed,
    setSideCollapsed: setCollapsed,
    mobileOpen: mobileOpen,
    setMobileOpen: setMobileOpen
  }, side && /*#__PURE__*/_react.default.createElement(_Nav2.default, _extends({
    appName: top ? '' : appName,
    hideCollapse: hideCollapse,
    collapsedAppName: collapsedAppName,
    logoHref: logoHref,
    logoOnClick: logoOnClick
  }, sideProps, {
    collapsed: collapsed,
    onCollapsedChange: setCollapsed,
    className: (0, _classnames.default)('from-xl', sideProps === null || sideProps === void 0 ? void 0 : sideProps.className)
  }), side, !top && !hideTheme && /*#__PURE__*/_react.default.createElement(_Nav8.default, null)), /*#__PURE__*/_react.default.createElement(_Nav.default, _extends({
    appName: appName,
    preLogo: (mobile || side) && /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      className: "to-xl",
      onClick: function onClick(e) {
        setMobileOpen(!mobileOpen);
        e.stopPropagation();
      }
    }, /*#__PURE__*/_react.default.createElement(_Nav4.default, {
      icon: mobileOpen ? _freeSolidSvgIcons.faTimes : _freeSolidSvgIcons.faBars
    })),
    logoHref: logoHref,
    logoOnClick: logoOnClick,
    hideTheme: hideTheme
  }, topProps, {
    onClick: function onClick(e) {
      var _topProps$onClick;

      setMobileOpen(false);
      topProps === null || topProps === void 0 ? void 0 : (_topProps$onClick = topProps.onClick) === null || _topProps$onClick === void 0 ? void 0 : _topProps$onClick.call(topProps, e);
    },
    className: (0, _classnames.default)(topProps === null || topProps === void 0 ? void 0 : topProps.className, {
      'to-xl': !top
    }),
    logoClassName: (0, _classnames.default)(topProps === null || topProps === void 0 ? void 0 : topProps.logoClassName, {
      'to-xl': side
    })
  }), top), (mobile || side) && /*#__PURE__*/_react.default.createElement(_Nav3.default, _extends({}, mobileProps, {
    onClick: function onClick(e) {
      var _mobileProps$onClick;

      // automatically close mobile nav on link clicks
      if (e.target.closest('a')) {
        setMobileOpen(false);
      }

      mobileProps === null || mobileProps === void 0 ? void 0 : (_mobileProps$onClick = mobileProps.onClick) === null || _mobileProps$onClick === void 0 ? void 0 : _mobileProps$onClick.call(mobileProps, e);
    },
    onOverlayClick: function onOverlayClick(e) {
      setMobileOpen(false);
      e.stopPropagation();
    },
    className: (0, _classnames.default)(mobileProps === null || mobileProps === void 0 ? void 0 : mobileProps.className, // for standalone <Mobile.Nav>, display the component by default,
    // but we want to hide it when using <Nav>
    'to-xl bf-display-none bf-scrollbar', {
      // only apply animation after first toggle to
      // prevent animation onload (overrides bf-display-none)
      'bf-nav-mobile-open': mobileOpen && hasToggledNavMobile,
      'bf-nav-mobile-close': !mobileOpen && hasToggledNavMobile
    })
  }), mobile !== null && mobile !== void 0 ? mobile : side), /*#__PURE__*/_react.default.createElement("main", _extends({
    ref: ref
  }, props), children));
});
Nav.Item = _Nav4.default;
Nav.Group = _Nav5.default;
Nav.Side = _Nav2.default;
Nav.Top = _Nav.default;
Nav.Mobile = _Nav3.default;
Nav.Header = _Nav6.default;
Nav.Search = _Nav7.default;
Nav.propTypes = {
  /** Your application content */
  children: _propTypes.default.node,

  /** Top nav content */
  top: _propTypes.default.node,

  /** Top bar props */
  topProps: _propTypes.default.object,

  /** Side nav content */
  side: _propTypes.default.node,

  /** Side nav props */
  sideProps: _propTypes.default.object,

  /** Mobile fullscreen nav content (optional, defaults to side content) */
  mobile: _propTypes.default.node,

  /** Mobile nav props */
  mobileProps: _propTypes.default.object,

  /** Application name (only use if no TopBar) */
  appName: _propTypes.default.node.isRequired,

  /** Abbreviated application name (max 4 characters, only use if no TopBar and you dont like the auto-generated one) */
  collapsedAppName: _propTypes.default.string,

  /** Hide collapse button */
  hideCollapse: _propTypes.default.bool,

  /** Hide the color scheme picker */
  hideTheme: _propTypes.default.bool,

  /** Custom url for clickable logo (default '/') */
  logoHref: _propTypes.default.string,

  /** Clickable logo event */
  logoOnClick: _propTypes.default.func
};
Nav.defaultProps = {
  logoHref: '/'
};
Nav.displayName = 'Nav';
var _default = Nav;
exports.default = _default;
