// @intility/bifrost-react/dist/hooks/useTheme.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ThemeContext = exports.updateBodyClassList = exports.usePreferredColorScheme = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ = require("..");

var _useMediaQuery = _interopRequireDefault(require("./useMediaQuery"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var themes = ['dark', 'light', 'system'];

var usePreferredColorScheme = function usePreferredColorScheme() {
  var systemDark = (0, _useMediaQuery.default)('(prefers-color-scheme: dark)'); // ssr or no way of knowing, return default behaviour

  if (systemDark === undefined) {
    return 'system';
  }

  return systemDark ? 'dark' : 'light';
}; //#region localStorageTheme
// Used as a fallback for when Bifrost are not present in consumer application


exports.usePreferredColorScheme = usePreferredColorScheme;
var subscribers = []; //subscribe func that returns unsubscribe func, this fits well with useEffect

var subscribe = function subscribe(fn) {
  subscribers.push(fn);
  return function () {
    subscribers = subscribers.filter(function (s) {
      return s !== fn;
    });
  };
}; //notify func that calls all subscribers


var notify = function notify(theme) {
  return subscribers.forEach(function (fn) {
    return fn(theme);
  });
};

var updateBodyClassList = function updateBodyClassList(theme) {
  //update body classes
  if (typeof document !== 'undefined') {
    document.body.classList.toggle('bf-lightmode', theme === 'light');
    document.body.classList.toggle('bf-darkmode', theme === 'dark');
  }
};

exports.updateBodyClassList = updateBodyClassList;

var updateTheme = function updateTheme(theme) {
  //update localStorage
  if (typeof window !== 'undefined' && window.localStorage) {
    localStorage.bfTheme = theme;
  }

  updateBodyClassList(theme); //update all hooks

  notify(theme);
}; //this subscribes with its setTheme function
//and exposes a toggle funciton that also notifies all other hooks


var useLocalStorageTheme = function useLocalStorageTheme(shouldModifyBody) {
  var startingTheme = typeof window !== 'undefined' && window.localStorage && themes.includes(localStorage.bfTheme) ? localStorage.bfTheme : 'system';

  var _useState = (0, _react.useState)(startingTheme),
      _useState2 = _slicedToArray(_useState, 2),
      theme = _useState2[0],
      setTheme = _useState2[1];

  var preferred = usePreferredColorScheme();
  (0, _react.useEffect)(function () {
    return subscribe(setTheme);
  }, []); // on load, set theme from localStorage

  (0, _.useIsomorphicLayoutEffect)(function () {
    if (shouldModifyBody) {
      updateBodyClassList(theme === 'system' ? preferred : theme);
    }
  }, [shouldModifyBody]); // Updates body class based on user preference if theme is set to system

  (0, _.useIsomorphicLayoutEffect)(function () {
    if (shouldModifyBody && theme === 'system') {
      updateBodyClassList(preferred);
    }
  }, [shouldModifyBody, theme, preferred]);
  return {
    theme: theme,
    setTheme: function setTheme(t) {
      updateTheme(t);
    },
    current: theme !== 'system' ? theme : preferred !== 'system' ? preferred : 'dark'
  };
}; //#endregion
//#region Context Theme
// Used when <Bifrost> wraps application


var ThemeContext = /*#__PURE__*/_react.default.createContext(null);

exports.ThemeContext = ThemeContext;
ThemeContext.displayName = 'BifrostThemeContext';

//#endregion
var useTheme = function useTheme() {
  var themeContext = (0, _react.useContext)(ThemeContext);
  var localStorageTheme = useLocalStorageTheme(!themeContext); // fall back to local storage when used without context

  return themeContext !== null && themeContext !== void 0 ? themeContext : localStorageTheme;
};

var _default = useTheme;
exports.default = _default;
