// @intility/bifrost-react/dist/components/Utility/Bifrost/Bifrost.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _useTheme = require("../../../hooks/useTheme");

var _useLocale = require("../../../hooks/useLocale");

var _ = require("../../..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * Configure Bifrost globally by wrapping this component around your app
 */
var Bifrost = function Bifrost(_ref) {
  var children = _ref.children,
      themeProp = _ref.theme,
      onThemeChange = _ref.onThemeChange,
      _ref$locale = _ref.locale,
      locale = _ref$locale === void 0 ? _.enUS : _ref$locale;

  var _useLocalStorage = (0, _.useLocalStorage)('bfTheme', themeProp !== null && themeProp !== void 0 ? themeProp : 'system'),
      _useLocalStorage2 = _slicedToArray(_useLocalStorage, 2),
      themeState = _useLocalStorage2[0],
      setTheme = _useLocalStorage2[1];

  var preferred = (0, _useTheme.usePreferredColorScheme)();
  var theme = themeProp !== null && themeProp !== void 0 ? themeProp : themeState;
  var handleThemeChange = onThemeChange !== null && onThemeChange !== void 0 ? onThemeChange : setTheme;
  var currentTheme = theme !== 'system' ? theme : preferred !== 'system' ? preferred : 'dark'; // apply css classes to body so CSS vars are available globally (body scrollbar, tippy popups)

  (0, _.useIsomorphicLayoutEffect)(function () {
    (0, _useTheme.updateBodyClassList)(currentTheme);
  }, [currentTheme]);
  return /*#__PURE__*/_react.default.createElement(_useLocale.LocaleContext.Provider, {
    value: locale
  }, /*#__PURE__*/_react.default.createElement(_useTheme.ThemeContext.Provider, {
    value: {
      theme: theme,
      setTheme: handleThemeChange,
      current: currentTheme
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('bf-options', {
      'bf-darkmode': (theme === 'system' ? preferred : theme) === 'dark',
      'bf-lightmode': (theme === 'system' ? preferred : theme) === 'light'
    })
  }, children)));
};

Bifrost.propTypes = {
  /** Your entire app (or at least the parts using bifrost components) */
  children: _propTypes.default.node,

  /** Current Bifrost theme: 'dark', 'light', or 'system' (default) */
  theme: _propTypes.default.oneOf(['dark', 'light', 'system']),

  /** Callback for theme changes */
  onThemeChange: _propTypes.default.func,

  /** Set current locale for all descendant Bifrost components */
  locale: _propTypes.default.object
};
var _default = Bifrost;
exports.default = _default;
