// @intility/bifrost-react/dist/components/Interactive/Modal/Modal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactModal = _interopRequireDefault(require("react-modal"));

var _Icon = _interopRequireDefault(require("../../Visual/Icon"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _faTimes = require("../../../assets/faTimes");

var _Modal = _interopRequireDefault(require("./Modal.Footer"));

var _Banner = _interopRequireDefault(require("../../Visual/Banner"));

var _useTheme2 = _interopRequireDefault(require("../../../hooks/useTheme"));

var _CloseButton = _interopRequireDefault(require("../common/CloseButton.internal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Show a Modal
 */
var Modal = function Modal(_ref) {
  var isOpen = _ref.isOpen,
      className = _ref.className,
      overlayClassName = _ref.overlayClassName,
      onRequestClose = _ref.onRequestClose,
      aria = _ref.aria,
      children = _ref.children,
      header = _ref.header,
      icon = _ref.icon,
      headerChildren = _ref.headerChildren,
      contentLabel = _ref.contentLabel,
      noPadding = _ref.noPadding,
      noHeader = _ref.noHeader,
      style = _ref.style,
      props = _objectWithoutProperties(_ref, ["isOpen", "className", "overlayClassName", "onRequestClose", "aria", "children", "header", "icon", "headerChildren", "contentLabel", "noPadding", "noHeader", "style"]);

  var _useTheme = (0, _useTheme2.default)(),
      current = _useTheme.current;

  return /*#__PURE__*/_react.default.createElement(_reactModal.default, _extends({
    isOpen: isOpen,
    className: "bf-modal bf-scrollbar",
    overlayClassName: (0, _classnames.default)("bf-".concat(current, "mode"), overlayClassName, 'bf-modal-wrapper'),
    onRequestClose: onRequestClose,
    contentLabel: noHeader ? header : contentLabel,
    style: style && {
      overlay: style.overlay
    },
    aria: contentLabel || noHeader ? aria : _objectSpread({
      labelledby: 'bf-modal-heading'
    }, aria)
  }, props), onRequestClose && /*#__PURE__*/_react.default.createElement(_Banner.default, {
    className: "bf-modal-close-button",
    onClick: onRequestClose
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _faTimes.faTimes,
    fixedWidth: true
  })), /*#__PURE__*/_react.default.createElement("div", {
    style: style && style.content,
    className: (0, _classnames.default)(className, 'bf-modal-content bf-scrollbar', {
      'bf-no-padding': noPadding
    })
  }, onRequestClose && /*#__PURE__*/_react.default.createElement(_CloseButton.default, {
    onClick: onRequestClose
  }), !noHeader && /*#__PURE__*/_react.default.createElement("header", {
    className: "bf-modal-header"
  }, /*#__PURE__*/_react.default.createElement("h1", {
    id: "bf-modal-heading",
    className: "bf-modal-header-h1"
  }, icon && /*#__PURE__*/_react.default.createElement(_Icon.default, {
    className: "bf-modal-header-icon",
    icon: icon
  }), header), headerChildren), /*#__PURE__*/_react.default.createElement("div", {
    className: "bf-modal-content-container"
  }, children)));
};

Modal.propTypes = {
  /** Boolean describing if the modal should be shown or not. */
  isOpen: _propTypes.default.bool.isRequired,

  /** Header text of the modal. This sets both contentLabel and aria.labelledby */
  header: _propTypes.default.string.isRequired,

  /** Header icon. See Icon component for more detail */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  icon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string]),

  /** Children of the header element, will be placed after the header text */
  headerChildren: _propTypes.default.node,

  /** CSS class name(s) for modal */
  className: _propTypes.default.string,

  /** Function that will be run when the modal is requested to be closed (either by clicking on overlay, clicking close button or pressing ESC). Note: It is not called if isOpen is changed by other means. */
  onRequestClose: _propTypes.default.func,

  /** Modal content */
  children: _propTypes.default.node,

  /** @ignore */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  aria: _propTypes.default.shape({
    labelledby: _propTypes.default.string,
    describedby: _propTypes.default.string,
    modal: _propTypes.default.oneOf([true, false, 'true', 'false'])
  }),

  /** @ignore */
  overlayClassName: _propTypes.default.string,

  /** @ignore */
  contentLabel: _propTypes.default.string,

  /** @ignore */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  style: _propTypes.default.shape({
    content: _propTypes.default.object,
    overlay: _propTypes.default.object
  }),

  /** Content should have no padding */
  noPadding: _propTypes.default.bool,

  /** Header should be visible */
  noHeader: _propTypes.default.bool
};
Modal.setAppElement = _reactModal.default.setAppElement;
Modal.Footer = _Modal.default;
Modal.displayName = 'Modal';
var _default = Modal;
exports.default = _default;
