// @intility/bifrost-react/dist/components/Interactive/Accordion/Accordion.Item.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _SlideDown = _interopRequireDefault(require("../../Utility/SlideDown"));

var _Icon = _interopRequireDefault(require("../../Visual/Icon"));

var _useAccordionContext2 = _interopRequireDefault(require("./hooks/useAccordionContext"));

var _faAngleRight = require("../../../assets/faAngleRight");

var _useLocale = _interopRequireDefault(require("../../../hooks/useLocale"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * A single pane for `<Accordion>` container
 * @visibleName Accordion.Item
 */
var AccordionItem = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var className = _ref.className,
      title = _ref.title,
      children = _ref.children,
      active = _ref.active,
      onClick = _ref.onClick,
      disabled = _ref.disabled,
      noPadding = _ref.noPadding,
      icon = _ref.icon,
      htmlTitle = _ref.htmlTitle,
      autoScroll = _ref.autoScroll,
      ariaLabel = _ref['aria-label'],
      props = _objectWithoutProperties(_ref, ["className", "title", "children", "active", "onClick", "disabled", "noPadding", "icon", "htmlTitle", "autoScroll", "aria-label"]);

  var locale = (0, _useLocale.default)();

  var _useAccordionContext = (0, _useAccordionContext2.default)(),
      onToggle = _useAccordionContext.onToggle,
      activeItem = _useAccordionContext.activeItem,
      variant = _useAccordionContext.variant;

  var isActive = active || activeItem === title;

  var handleChange = function handleChange(key) {
    if (onClick) {
      onClick(key);
    } else if (onToggle) {
      onToggle(key === activeItem ? null : key);
    }
  };

  return /*#__PURE__*/_react.default.createElement("div", _extends({
    className: (0, _classnames.default)(className, 'bf-accordion-item', {
      'bf-accordion-item-active': isActive,
      'bf-accordion-item-disabled': disabled
    }),
    "data-testid": "bf-accordion-item",
    ref: ref
  }, props), /*#__PURE__*/_react.default.createElement("button", {
    className: "bf-accordion-item-title",
    "data-testid": "bf-accordion-item-title",
    onClick: function onClick() {
      return handleChange(title);
    },
    title: htmlTitle !== null && htmlTitle !== void 0 ? htmlTitle : typeof title === 'string' ? title : undefined,
    "aria-label": ariaLabel,
    disabled: disabled,
    type: "button"
  }, variant !== 'styled' && /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _faAngleRight.faAngleRight,
    fixedWidth: true,
    className: "bf-accordion-item-angle"
  }), icon && /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: icon,
    fixedWidth: true,
    className: "bf-accordion-item-icon"
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-neutral-link-text"
  }, title), variant === 'styled' && /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-accordion-item-angle-wrapper"
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-neutral-link-text"
  }, isActive ? locale.hide : locale.show), /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _faAngleRight.faAngleRight,
    fixedWidth: true,
    className: "bf-accordion-item-angle"
  }))), /*#__PURE__*/_react.default.createElement(_SlideDown.default, {
    contentClassName: (0, _classnames.default)('bf-accordion-item-content', {
      'bfl-nopadding': noPadding
    }),
    open: isActive,
    autoScroll: autoScroll,
    "data-testid": "bf-accordion-item-content"
  }, children));
}); // todo: fix proptypes to match typescript, if possible? herman halp!

AccordionItem.propTypes = {
  /** CSS class name(s) */
  className: _propTypes.default.string,

  /** Title of accordion item. **Note:** it is important that this prop is unique (unless you supply your own active state and onChange handler) */
  title: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.element]).isRequired,

  /** Content to show when accordion item is open */
  children: _propTypes.default.node,

  /** Overrides default onClick on Accordion.Item */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  onClick: _propTypes.default.func,

  /** Bool to indicate when item should be opened */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  active: _propTypes.default.bool,

  /** Disabled accordion item */
  disabled: _propTypes.default.bool,

  /** Bool to remove padding from content div */
  noPadding: _propTypes.default.bool,

  /** Font Awesome icon reference (or string if using library) */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  icon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string]),

  /** Title to use for long titles cut off with ellipsis, not needed if title is string */
  htmlTitle: _propTypes.default.string,

  /** Automatically scroll to view expanded content */
  autoScroll: _propTypes.default.bool
};
AccordionItem.defaultProps = {
  disabled: false
};
AccordionItem.displayName = 'Accordion.Item';
var _default = AccordionItem;
exports.default = _default;
