// @intility/bifrost-react/es/components/Navigation/TabBar/TabBar.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { forwardRef, useRef } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import TabBarItem from './TabBar.Item';
import Button from '../../Inputs/Button';
import useOverflow from '../../Interactive/Tabs/hooks/useOverflow';
import { faAngleLeft, faAngleRight } from '../../../assets';
import useMediaQuery from '../../../hooks/useMediaQuery';

/**
 * TabBar component - switch between different content
 */
var TabBar = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var className = _ref.className,
      variant = _ref.variant,
      props = _objectWithoutProperties(_ref, ["className", "variant"]);

  var elRef = useRef(null);

  var _useOverflow = useOverflow(elRef),
      refOverflow = _useOverflow.refOverflow,
      refScrollBegin = _useOverflow.refScrollBegin,
      refScrollEnd = _useOverflow.refScrollEnd;

  var isWide = useMediaQuery('(min-width: 600px)');

  var handleArrowClick = function handleArrowClick(direction) {
    var elementRef = elRef.current;
    if (!elementRef) return;

    if (direction === 'left') {
      elementRef.scrollLeft -= elementRef.clientWidth / 2;
    } else {
      elementRef.scrollLeft += elementRef.clientWidth / 2;
    }
  };

  var scrollToTab = function scrollToTab(e) {
    var tab = e.target.closest('.bf-tab');

    if (tab) {
      if (!elRef.current) return;
      var container = elRef.current;
      var padding = isWide ? 28 : 16;
      var containerLeftSide = container.scrollLeft + padding; // first tab starts left on 28px

      var tabLeftSide = tab.offsetLeft;
      var containerRightSide = container.scrollLeft + container.offsetWidth;
      var tabRightSide = tabLeftSide + tab.clientWidth; // Right side of tab is out of view

      if (tabRightSide > containerRightSide - 15) {
        var scrollValue = tabRightSide - containerRightSide - 12 + (isWide ? 0 : 26) + (variant ? 4 : 0);
        container.scrollLeft += scrollValue;
      } // Left side of tab is out of view
      else if (tabLeftSide < containerLeftSide + 15) {
          var _scrollValue = containerLeftSide - tabLeftSide + 10 + (isWide ? 0 : 16) + (variant ? 0 : 0);

          container.scrollLeft -= _scrollValue;
        }
    }
  };

  return /*#__PURE__*/React.createElement("div", _extends({
    className: classnames('bf-tabs-container', {
      'bf-tabs-styled': variant === 'styled'
    }),
    ref: ref
  }, props), /*#__PURE__*/React.createElement("div", {
    ref: elRef,
    className: classnames('bf-tabs', className),
    "data-testid": "bf-tabs",
    onClick: function onClick(e) {
      return scrollToTab(e);
    }
  }, props.children), /*#__PURE__*/React.createElement(Button, {
    className: classnames('bf-tab-left', {
      'bf-tab-arrow-hidden': !refOverflow || refScrollBegin
    }),
    onClick: function onClick() {
      return handleArrowClick('left');
    },
    icon: faAngleLeft,
    tabIndex: -1
  }), /*#__PURE__*/React.createElement(Button, {
    className: classnames('bf-tab-right', {
      'bf-tab-arrow-hidden': !refOverflow || refScrollEnd
    }),
    onClick: function onClick() {
      return handleArrowClick('right');
    },
    icon: faAngleRight,
    tabIndex: -1
  }));
});
TabBar.propTypes = {
  /** CSS classname(s) */
  className: PropTypes.string,

  /** Collection of TabBar.Item components */
  children: PropTypes.node,

  /** Can be used to change the menu's appearance. Available variants: 'default' (default) and 'styled' */
  variant: PropTypes.oneOf(['default', 'styled'])
};
TabBar.defaultProps = {
  variant: 'default'
};
TabBar.Item = TabBarItem;
TabBar.displayName = 'TabBar';
export default TabBar;
