// @intility/bifrost-react/dist/components/Interactive/Pagination/Pagination.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = _interopRequireDefault(require("../../Visual/Icon"));

var _assets = require("../../../assets");

var _useLocale = _interopRequireDefault(require("../../../hooks/useLocale"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Component for displaying available pages and current page in a paged environment
 */
var Pagination = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var className = _ref.className,
      currentPage = _ref.currentPage,
      totalPages = _ref.totalPages,
      _ref$displayPages = _ref.displayPages,
      displayPages = _ref$displayPages === void 0 ? 9 : _ref$displayPages,
      onChange = _ref.onChange,
      props = _objectWithoutProperties(_ref, ["className", "currentPage", "totalPages", "displayPages", "onChange"]);

  var locale = (0, _useLocale.default)();
  var displayPagesCapForMediumScreens = Math.min(displayPages, 9);
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    className: (0, _classnames.default)(className, 'bf-pagination'),
    ref: ref,
    "data-testid": "bf-pagination"
  }, props), currentPage > 1 && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(locale.goTo, " ").concat(locale.first, " ").concat(locale.page),
    type: "button",
    className: "bf-pagination-button bf-pagination-button-arrow to-small",
    onClick: function onClick() {
      return onChange(1);
    }
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _assets.faAngleDoubleLeft
  })), /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(locale.goTo, " ").concat(locale.previous, " ").concat(locale.page),
    className: "bf-title-link bf-pagination-prev",
    onClick: function onClick() {
      return onChange(currentPage - 1);
    }
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _assets.faAngleLeft
  }), locale.prev)), /*#__PURE__*/_react.default.createElement("span", {
    className: "to-small"
  }, "".concat(currentPage, " of ").concat(totalPages)), /*#__PURE__*/_react.default.createElement("span", {
    className: "from-small to-medium"
  }, totalPages > displayPagesCapForMediumScreens ? /*#__PURE__*/_react.default.createElement(ComplexPagination, {
    currentPage: currentPage,
    onChange: onChange,
    displayPages: displayPagesCapForMediumScreens,
    totalPages: totalPages
  }) : /*#__PURE__*/_react.default.createElement(BasicPagination, {
    currentPage: currentPage,
    onChange: onChange,
    totalPages: totalPages
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: "from-medium"
  }, totalPages > displayPages ? /*#__PURE__*/_react.default.createElement(ComplexPagination, {
    currentPage: currentPage,
    onChange: onChange,
    displayPages: displayPages,
    totalPages: totalPages
  }) : /*#__PURE__*/_react.default.createElement(BasicPagination, {
    currentPage: currentPage,
    onChange: onChange,
    totalPages: totalPages
  })), currentPage < totalPages && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(locale.goTo, " ").concat(locale.next, " ").concat(locale.page),
    className: "bf-title-link bf-pagination-next",
    onClick: function onClick() {
      return onChange(currentPage + 1);
    }
  }, locale.next, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _assets.faAngleRight
  })), /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(locale.goTo, " ").concat(locale.last, " ").concat(locale.page),
    type: "button",
    className: "bf-pagination-button bf-pagination-button-arrow to-small",
    onClick: function onClick() {
      return onChange(totalPages);
    }
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _assets.faAngleDoubleRight
  }))));
});
Pagination.propTypes = {
  /** CSS class name(s) */
  className: _propTypes.default.string,

  /** Current page */
  currentPage: _propTypes.default.number.isRequired,

  /** Total number of pages */
  totalPages: _propTypes.default.number.isRequired,

  /** Callback for page change. Calls the function with the page number */
  onChange: _propTypes.default.func.isRequired,

  /** Total number of displayed pages (includes '...'), should be odd and 7 or higher */
  displayPages: _propTypes.default.number
};
Pagination.defaultProps = {
  displayPages: 9
}; // INTERNAL PageButton

var PageButton = function PageButton(_ref2) {
  var page = _ref2.page,
      currentPage = _ref2.currentPage,
      onChange = _ref2.onChange;
  var locale = (0, _useLocale.default)();
  return /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(locale.goTo, " ").concat(locale.page, " ").concat(page),
    type: "button",
    disabled: page === currentPage,
    className: (0, _classnames.default)('bf-pagination-button', {
      'bf-pagination-button-selected': page === currentPage
    }),
    onClick: function onClick() {
      return onChange(page);
    }
  }, page);
};

PageButton.propTypes = {
  page: _propTypes.default.number.isRequired,
  currentPage: _propTypes.default.number.isRequired,
  onChange: _propTypes.default.func.isRequired
}; // INTERNAL BasicPagination

var BasicPagination = function BasicPagination(_ref3) {
  var currentPage = _ref3.currentPage,
      totalPages = _ref3.totalPages,
      onChange = _ref3.onChange,
      _ref3$offset = _ref3.offset,
      offset = _ref3$offset === void 0 ? 1 : _ref3$offset;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, Array(totalPages).fill(null).map(function (_, i) {
    return /*#__PURE__*/_react.default.createElement(PageButton, {
      key: i + offset,
      page: i + offset,
      currentPage: currentPage,
      onChange: onChange
    });
  }));
};

BasicPagination.propTypes = {
  currentPage: _propTypes.default.number.isRequired,
  totalPages: _propTypes.default.number.isRequired,
  onChange: _propTypes.default.func.isRequired,
  offset: _propTypes.default.number
};
BasicPagination.defaultProps = {
  offset: 1
}; // INTERNAL ComplexPagination

var ComplexPagination = function ComplexPagination(_ref4) {
  var currentPage = _ref4.currentPage,
      totalPages = _ref4.totalPages,
      _ref4$displayPages = _ref4.displayPages,
      dp = _ref4$displayPages === void 0 ? 9 : _ref4$displayPages,
      onChange = _ref4.onChange;
  var displayPages = Math.max(7, dp % 2 === 0 ? dp + 1 : dp);
  var dynamicPages = displayPages - 4;
  var dynamicOffset = Math.min(Math.max(currentPage - (displayPages - 5) / 2, 3), totalPages - 1 - dynamicPages);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(PageButton, {
    page: 1,
    currentPage: currentPage,
    onChange: onChange
  }), currentPage <= Math.ceil(displayPages / 2) ? /*#__PURE__*/_react.default.createElement(PageButton, {
    page: 2,
    currentPage: currentPage,
    onChange: onChange
  }) : /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-pagination-dots"
  }, "..."), /*#__PURE__*/_react.default.createElement(BasicPagination, {
    totalPages: dynamicPages,
    currentPage: currentPage,
    offset: dynamicOffset,
    onChange: onChange
  }), currentPage >= totalPages - Math.floor(displayPages / 2) ? /*#__PURE__*/_react.default.createElement(PageButton, {
    page: totalPages - 1,
    currentPage: currentPage,
    onChange: onChange
  }) : /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-pagination-dots"
  }, "..."), /*#__PURE__*/_react.default.createElement(PageButton, {
    page: totalPages,
    currentPage: currentPage,
    onChange: onChange
  }));
};

ComplexPagination.propTypes = {
  currentPage: _propTypes.default.number.isRequired,
  totalPages: _propTypes.default.number.isRequired,
  onChange: _propTypes.default.func.isRequired,
  displayPages: _propTypes.default.number.isRequired
};
Pagination.displayName = 'Pagination';
var _default = Pagination;
exports.default = _default;
