// @intility/bifrost-react/dist/components/Inputs/TextArea/TextArea.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _Icon = _interopRequireDefault(require("../../Visual/Icon"));

var _common = require("../common");

var _useUniqueId = _interopRequireDefault(require("../../../hooks/useUniqueId"));

var _Label = _interopRequireDefault(require("../Label"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * TextArea field
 */
var TextArea = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var label = _ref.label,
      className = _ref.className,
      id = _ref.id,
      disabled = _ref.disabled,
      state = _ref.state,
      feedback = _ref.feedback,
      required = _ref.required,
      hideLabel = _ref.hideLabel,
      description = _ref.description,
      requiredNoLabel = _ref.requiredNoLabel,
      optional = _ref.optional,
      containerProps = _ref.containerProps,
      props = _objectWithoutProperties(_ref, ["label", "className", "id", "disabled", "state", "feedback", "required", "hideLabel", "description", "requiredNoLabel", "optional", "containerProps"]);

  var textareaId = (0, _useUniqueId.default)(id);
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    "data-testid": "bf-textarea-container"
  }, containerProps, {
    /* spread first so we can override className */
    className: (0, _classnames.default)(containerProps && containerProps.className, {
      'bf-textarea-container': true
    }, {
      'bf-textarea-disabled': disabled
    }, {
      'bf-textarea-warning': state === 'warning'
    }, {
      'bf-textarea-alert': state === 'alert'
    })
  }), !hideLabel && /*#__PURE__*/_react.default.createElement(_Label.default, {
    htmlFor: textareaId,
    required: required && !requiredNoLabel,
    optional: optional,
    disabled: disabled
  }, label), description && /*#__PURE__*/_react.default.createElement(_common.Description, null, description), /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": "bf-textarea-icon-container",
    className: "bf-textarea-icon-container"
  }, /*#__PURE__*/_react.default.createElement("textarea", _extends({
    id: textareaId,
    "data-testid": "bf-textarea",
    className: (0, _classnames.default)('bf-textarea', 'bf-scrollbar', className),
    disabled: disabled
  }, hideLabel && {
    'aria-label': label
  }, {
    required: !disabled && (required || requiredNoLabel),
    ref: ref
  }, props)), !disabled && state && (state === 'warning' || state === 'alert') && /*#__PURE__*/_react.default.createElement(_Icon.default, {
    className: "bf-textarea-icon",
    icon: _freeSolidSvgIcons.faExclamationTriangle
  })), feedback && /*#__PURE__*/_react.default.createElement(_common.Feedback, null, feedback));
});
TextArea.propTypes = {
  /** Required label text */
  label: _propTypes.default.string.isRequired,

  /** CSS class name(s) */
  className: _propTypes.default.string,

  /** Unique id for input and label */
  id: _propTypes.default.string,

  /** Disables the textarea */
  disabled: _propTypes.default.bool,

  /** Available states: `theme` (default), `warning`, and `alert` */
  state: _propTypes.default.oneOf(['warning', 'alert']),

  /** Feedback text below the textarea */
  feedback: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),

  /** Required prop indicating mandatory field */
  required: _propTypes.default.bool,

  /** Bool to hide label */
  hideLabel: _propTypes.default.bool,

  /** Description text to show more information */
  description: _propTypes.default.string,

  /** Bool to mark textarea required, but hides the required label */
  requiredNoLabel: _propTypes.default.bool,

  /** Marks textarea as optional */
  optional: _propTypes.default.bool,

  /** Props for the containing `div` element */
  containerProps: _propTypes.default.object
};
TextArea.defaultProps = {
  disabled: false,
  required: false,
  requiredNoLabel: false,
  optional: false
};
TextArea.displayName = 'TextArea';
var _default = TextArea;
exports.default = _default;
