// @intility/bifrost-react/es/components/Navigation/Nav/Nav.Mobile.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import NavLocationProvider from './NavLocationContext.internal';

/**
 * Full-screen fixed mobile nav
 */
var NavMobile = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var children = _ref.children,
      className = _ref.className,
      onOverlayClick = _ref.onOverlayClick,
      props = _objectWithoutProperties(_ref, ["children", "className", "onOverlayClick"]);

  return /*#__PURE__*/React.createElement(NavLocationProvider, {
    where: "mobile",
    open: true
  }, /*#__PURE__*/React.createElement("nav", _extends({}, props, {
    className: classnames('bf-nav bf-nav-mobile', className),
    ref: ref
  }), children), onOverlayClick && /*#__PURE__*/React.createElement("div", {
    className: classnames('bf-nav-mobile-overlay', className),
    onClick: onOverlayClick
  }));
});
NavMobile.propTypes = {
  /** Mobile nav content */
  children: PropTypes.node,

  /** CSS class name(s) for both nav and overlay (if any) */
  className: PropTypes.string,

  /** Callback for overlay click. Overlay is only rendered if a callback is supplied */
  onOverlayClick: PropTypes.func
};
NavMobile.displayName = 'Nav.Mobile';
export default NavMobile;
