// @intility/bifrost-react/es/components/Navigation/Nav/Nav.Group.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useState, useEffect, useCallback, useRef, forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { hideAll } from 'tippy.js';
import Expander from './Expander.internal';
import { Icon } from '../../..';
import { faAngleUp } from '../../../assets/faAngleUp';
import { faAngleDown } from '../../../assets/faAngleDown';
import NavItem from './Nav.Item';
import { useNavLocation } from './NavLocationContext.internal';
import useIsomorphicLayoutEffect from '../../../hooks/useIsomorphicLayoutEffect';
import Dropdown from '../../Interactive/Dropdown';

// https://gist.github.com/dandelany/1ff06f4fa1f8d6f89c5e#gistcomment-3090249
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function recursiveCloneChildren(children, newProps) {
  return React.Children.map(children, function (child) {
    if (! /*#__PURE__*/React.isValidElement(child)) {
      return child;
    } // Eg. String has no props


    if (child.props) {
      newProps.children = recursiveCloneChildren(child.props.children, newProps);
      return /*#__PURE__*/React.cloneElement(child, newProps);
    }

    return child;
  });
}
/**
 * Group of clickable items for Nav
 */


var NavGroup = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var _tippyRef$current5;

  var active = _ref.active,
      name = _ref.name,
      icon = _ref.icon,
      children = _ref.children,
      className = _ref.className,
      autoScroll = _ref.autoScroll,
      props = _objectWithoutProperties(_ref, ["active", "name", "icon", "children", "className", "autoScroll"]);

  var _useNavLocation = useNavLocation(),
      where = _useNavLocation.where,
      open = _useNavLocation.open;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      expanded = _useState2[0],
      setExpanded = _useState2[1];

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      tippyVisible = _useState4[0],
      setTippyVisible = _useState4[1];

  var buttonRef = useRef(null);
  var tippyRef = useRef();
  var childrenRef = useRef(null); // if active prop is undefined, set active class on group if any child <a> or <button> DOM node has an active class

  useIsomorphicLayoutEffect(function () {
    if (childrenRef.current && buttonRef.current && active === undefined) {
      buttonRef.current.classList.toggle('active', childrenRef.current.querySelectorAll('a.active, button.active').length > 0);
    }
  });

  var closePopup = function closePopup() {
    var _tippyRef$current, _tippyRef$current$hid, _focus, _ref2, _tippyRef$current2;

    (_tippyRef$current = tippyRef.current) === null || _tippyRef$current === void 0 ? void 0 : (_tippyRef$current$hid = _tippyRef$current.hide) === null || _tippyRef$current$hid === void 0 ? void 0 : _tippyRef$current$hid.call(_tippyRef$current);
    (_focus = (_ref2 = (_tippyRef$current2 = tippyRef.current) === null || _tippyRef$current2 === void 0 ? void 0 : _tippyRef$current2.reference).focus) === null || _focus === void 0 ? void 0 : _focus.call(_ref2);
  };

  var togglePopup = function togglePopup() {
    var _tippyRef$current3;

    if ((_tippyRef$current3 = tippyRef.current) !== null && _tippyRef$current3 !== void 0 && _tippyRef$current3.state.isShown) {
      closePopup();
    } else {
      var _tippyRef$current4;

      (_tippyRef$current4 = tippyRef.current) === null || _tippyRef$current4 === void 0 ? void 0 : _tippyRef$current4.show();
    }
  };

  useEffect(function () {
    if (open) hideAll();
  }, [open]);
  var handleClick = useCallback(function (e) {
    e.preventDefault();

    if (!open) {
      togglePopup();
    } else {
      setExpanded(!expanded);
    }
  }, [setExpanded, expanded, open]);
  return /*#__PURE__*/React.createElement("div", _extends({}, props, {
    ref: ref,
    className: classNames('bf-nav-group', className)
  }), /*#__PURE__*/React.createElement("button", {
    type: "button",
    className: classNames('bf-nav-group-button', {
      'bf-nav-group-button-expanded': expanded,
      'bf-nav-group-dropout-visible': tippyVisible,
      active: active
    }),
    "aria-expanded": !open ? (_tippyRef$current5 = tippyRef.current) === null || _tippyRef$current5 === void 0 ? void 0 : _tippyRef$current5.state.isShown : expanded,
    onClick: handleClick,
    ref: buttonRef
  }, /*#__PURE__*/React.createElement(NavItem, {
    icon: icon
  }, name, open && /*#__PURE__*/React.createElement("span", {
    className: "bf-nav-expand-icon"
  }, /*#__PURE__*/React.createElement(Icon, {
    icon: expanded ? faAngleUp : faAngleDown,
    fixedWidth: true
  })))), (where === 'top' || where === 'side' && !open) && /*#__PURE__*/React.createElement(Dropdown, {
    placement: where === 'top' ? 'bottom-end' : 'right-start',
    onCreate: function onCreate(instance) {
      return tippyRef.current = instance;
    },
    onShow: function onShow() {
      setTippyVisible(true);
      hideAll();
    },
    onHide: function onHide() {
      return setTippyVisible(false);
    },
    className: "bf-nav-group-dropout",
    popperOptions: {
      strategy: 'fixed'
    },
    reference: buttonRef,
    content: /*#__PURE__*/React.createElement("div", {
      className: "bf-nav-group-dropout-content",
      onClick: function onClick(e) {
        if (e.target.closest('.bf-nav-item')) {
          closePopup();
        }
      }
    }, where === 'side' && name ? /*#__PURE__*/React.createElement("div", {
      className: "bf-nav-group-dropout-title"
    }, name) : null, children)
  }), where !== 'top' && /*#__PURE__*/React.createElement(Expander, {
    expanded: expanded,
    autoScroll: autoScroll
  }, children), /*#__PURE__*/React.createElement("div", {
    style: {
      display: 'none'
    },
    ref: childrenRef
  }, recursiveCloneChildren(children, {
    id: undefined,
    htmlFor: undefined,
    name: undefined
  })));
});
NavGroup.propTypes = {
  /** Group button label */
  name: PropTypes.node,

  /** CSS class name(s) */
  className: PropTypes.string,

  /** Nav group content (if not a simple string, also supply text version as title) */
  children: PropTypes.node,

  /** Set active to true when a the current page is a sublink of this group */
  active: PropTypes.bool,

  /** Font Awesome icon reference (or string if using library) */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  icon: PropTypes.oneOfType([PropTypes.object, PropTypes.array, PropTypes.string]),

  /** Automatically scroll to view expanded content */
  autoScroll: PropTypes.bool
};
NavGroup.displayName = 'Nav.Group';
export default NavGroup;
