// @intility/bifrost-react/dist/components/Inputs/Input/Input.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _classnames = _interopRequireDefault(require("classnames"));

var _types = require("../../../types");

var _Icon = _interopRequireDefault(require("../../Visual/Icon"));

var _setRef = _interopRequireDefault(require("../../../utils/setRef"));

var _triggerOnChange = _interopRequireDefault(require("../../../utils/triggerOnChange"));

var _faTimes = require("../../../assets/faTimes");

var _common = require("../common");

var _useUniqueId = _interopRequireDefault(require("../../../hooks/useUniqueId"));

var _Label = _interopRequireDefault(require("../Label"));

var _assets = require("../../../assets");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Text input field
 */
var Input = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, _ref2) {
  var _inputRef$current;

  var label = _ref.label,
      hideLabel = _ref.hideLabel,
      icon = _ref.icon,
      iconProps = _ref.iconProps,
      rightIcon = _ref.rightIcon,
      className = _ref.className,
      id = _ref.id,
      disabled = _ref.disabled,
      state = _ref.state,
      feedback = _ref.feedback,
      required = _ref.required,
      description = _ref.description,
      requiredNoLabel = _ref.requiredNoLabel,
      optional = _ref.optional,
      variant = _ref.variant,
      clearable = _ref.clearable,
      loading = _ref.loading,
      onIconClick = _ref.onIconClick,
      iconButton = _ref.iconButton,
      containerProps = _ref.containerProps,
      props = _objectWithoutProperties(_ref, ["label", "hideLabel", "icon", "iconProps", "rightIcon", "className", "id", "disabled", "state", "feedback", "required", "description", "requiredNoLabel", "optional", "variant", "clearable", "loading", "onIconClick", "iconButton", "containerProps"]);

  var inputId = (0, _useUniqueId.default)(id);
  var isOutline = variant === 'outline';
  var hasState = !!state && state !== 'theme';
  var hasIcon = !!icon || !!(iconProps !== null && iconProps !== void 0 && iconProps.icon);
  var inputRef = (0, _react.useRef)(null);

  var handleClear = function handleClear() {
    if (disabled) return;

    if (inputRef.current) {
      (0, _triggerOnChange.default)(inputRef.current, '');
      inputRef.current.focus();
    }
  };

  var inputIcon = null;

  if (hasIcon && !(hasState && rightIcon)) {
    inputIcon = /*#__PURE__*/_react.default.createElement(_Icon.default // @ts-ignore
    , _extends({
      icon: loading ? _assets.fadSpinnerThird : icon,
      spin: loading,
      fixedWidth: true
    }, iconProps, {
      className: (0, _classnames.default)(iconProps === null || iconProps === void 0 ? void 0 : iconProps.className, {
        'bf-input-loading-icon': loading
      }),
      "data-testid": "bf-input-icon"
    }));
    inputIcon = onIconClick ? /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      className: "bf-input-icon bf-input-icon-button",
      onClick: onIconClick,
      disabled: disabled
    }, inputIcon) : /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _classnames.default)(iconProps === null || iconProps === void 0 ? void 0 : iconProps.className, 'bf-input-icon')
    }, inputIcon);
  } else if (loading) {
    inputIcon = /*#__PURE__*/_react.default.createElement("span", {
      className: "bf-input-icon"
    }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
      icon: _assets.fadSpinnerThird,
      fixedWidth: true,
      spin: true,
      className: "bf-input-loading-icon"
    }));
    rightIcon = true;
    hasState = false;
  }

  return /*#__PURE__*/_react.default.createElement("div", _extends({}, containerProps, {
    /* spread first so we can override className */
    className: (0, _classnames.default)('bf-input-container', containerProps === null || containerProps === void 0 ? void 0 : containerProps.className, {
      'bf-input-disabled': disabled,
      'bf-input-success': state === 'success' && !isOutline,
      'bf-input-warning': state === 'warning' && !isOutline,
      'bf-input-alert': state === 'alert' && !isOutline
    }),
    "data-testid": "bf-input-container"
  }), !hideLabel && /*#__PURE__*/_react.default.createElement(_Label.default, {
    htmlFor: inputId,
    required: !disabled && required && !requiredNoLabel,
    optional: optional,
    disabled: disabled
  }, label), description && /*#__PURE__*/_react.default.createElement(_common.Description, null, description), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('bf-input-icon-container', {
      'bf-input-icon-right': !hasState && inputIcon && rightIcon,
      'bf-input-icon-left': inputIcon && !rightIcon,
      'bf-input-icon-state': hasState && !isOutline,
      'bf-input-outline': isOutline,
      'bf-input-clearable': clearable,
      'bf-input-icon-button-filled': iconButton
    }),
    "data-testid": "bf-input-icon-container"
  }, /*#__PURE__*/_react.default.createElement("input", _extends({
    id: inputId,
    className: (0, _classnames.default)(className, 'bf-input', {
      'bf-input-state-only': hasState && !inputIcon
    }),
    disabled: disabled,
    required: !disabled && (required || requiredNoLabel)
  }, hideLabel && {
    'aria-label': label
  }, {
    ref: function ref(r) {
      (0, _setRef.default)(_ref2, r);
      (0, _setRef.default)(inputRef, r);
    },
    autoComplete: "noAutocomplete"
  }, props)), inputIcon, clearable && (props.value || ((_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.value)) && /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-input-clear-icon",
    onClick: function onClick() {
      return handleClear();
    }
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _faTimes.faTimes
  })), hasState && !isOutline && /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-state-icon"
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    fixedWidth: true,
    icon: state === 'alert' || state === 'warning' ? _freeSolidSvgIcons.faExclamationTriangle : _freeSolidSvgIcons.faCheck
  }))), typeof feedback === 'string' && /*#__PURE__*/_react.default.createElement(_common.Feedback, null, feedback || "\xA0"));
}); //@ts-ignore

Input.propTypes = {
  /** Required label text */
  label: _propTypes.default.string.isRequired,

  /** Hide label text, will add label prop as aria-label for screen readers */
  hideLabel: _propTypes.default.bool,

  /** Font Awesome icon reference (or string if using library) */
  //@ts-ignore
  icon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string]),

  /** Icon position. Default is left side */
  rightIcon: _propTypes.default.bool,

  /** CSS class name(s) */
  className: _propTypes.default.string,

  /** Unique id for input and label */
  id: _propTypes.default.string,

  /** Disables the input field */
  disabled: _propTypes.default.bool,

  /** Available states: `theme` (default), `success`, `warning`, and `alert` */
  state: _propTypes.default.oneOf(_types.StateArray),

  /** Feedback text below the input field */
  feedback: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),

  /** Required prop indicating mandatory field */
  required: _propTypes.default.bool,

  /** Description text to show more information */
  description: _propTypes.default.string,

  /** Bool to mark input required, but hides the required label */
  requiredNoLabel: _propTypes.default.bool,

  /** Marks input as optional */
  optional: _propTypes.default.bool,

  /** Available variants: 'basic' (default), and 'outline' */
  variant: _propTypes.default.oneOf(['basic', 'outline']),

  /** Display a clickable 'x' to clear the field */
  clearable: _propTypes.default.bool,

  /** Display a loading spinner in the input */
  loading: _propTypes.default.bool,

  /** @ignore */
  value: _propTypes.default.any,

  /** Props for the containing `div` element */
  containerProps: _propTypes.default.object,

  /** Icon component props */
  //@ts-ignore
  iconProps: _propTypes.default.shape({
    icon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string]),
    className: _propTypes.default.string,
    onClick: _propTypes.default.func
  }),

  /** onClick handler for icon */
  onIconClick: _propTypes.default.func,

  /** Display icon as filled button */
  iconButton: _propTypes.default.bool
};
Input.defaultProps = {
  variant: 'basic'
};
Input.displayName = 'Input';
var _default = Input;
exports.default = _default;
