// @intility/bifrost-react/dist/components/Navigation/TopBar/TopBar.Group.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react2 = _interopRequireDefault(require("@tippyjs/react"));

var _tippy = require("tippy.js");

var _TopBar = _interopRequireDefault(require("./TopBar.Item"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var TopBarGroup = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var name = _ref.name,
      children = _ref.children,
      className = _ref.className,
      style = _ref.style,
      icon = _ref.icon,
      rightIcon = _ref.rightIcon,
      selectable = _ref.selectable,
      props = _objectWithoutProperties(_ref, ["name", "children", "className", "style", "icon", "rightIcon", "selectable"]);

  var tippyRef = (0, _react.useRef)();

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      tippyVisible = _useState2[0],
      setTippyVisible = _useState2[1];

  var closePopup = function closePopup() {
    var _tippyRef$current, _focus, _ref2, _tippyRef$current2;

    (_tippyRef$current = tippyRef.current) === null || _tippyRef$current === void 0 ? void 0 : _tippyRef$current.hide();
    (_focus = (_ref2 = (_tippyRef$current2 = tippyRef.current) === null || _tippyRef$current2 === void 0 ? void 0 : _tippyRef$current2.reference).focus) === null || _focus === void 0 ? void 0 : _focus.call(_ref2);
  };

  var active = (Array.isArray(children) ? children : [children]).some(function (node) {
    var _node$props$className, _node$props$className2;

    return ((_node$props$className = (_node$props$className2 = node.props.className) === null || _node$props$className2 === void 0 ? void 0 : _node$props$className2.toLowerCase()) !== null && _node$props$className !== void 0 ? _node$props$className : '').includes('active');
  });

  var togglePopup = function togglePopup() {
    var _tippyRef$current3;

    if ((_tippyRef$current3 = tippyRef.current) !== null && _tippyRef$current3 !== void 0 && _tippyRef$current3.state.isShown) {
      closePopup();
    } else {
      var _tippyRef$current4;

      (_tippyRef$current4 = tippyRef.current) === null || _tippyRef$current4 === void 0 ? void 0 : _tippyRef$current4.show();
    }
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": "bf-topbar-group-container",
    className: (0, _classnames.default)(className, 'bf-topbar-group-container'),
    style: style,
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(_react2.default, {
    placement: "bottom-end",
    trigger: "mouseenter",
    arrow: true,
    onCreate: function onCreate(instance) {
      return tippyRef.current = instance;
    },
    interactive: true,
    offset: [0, 10],
    onShow: function onShow() {
      setTippyVisible(true);
      (0, _tippy.hideAll)();
    },
    onHide: function onHide() {
      return setTippyVisible(false);
    },
    className: "bf-topbar-group",
    content: /*#__PURE__*/_react.default.createElement("div", {
      onClick: closePopup
    }, children),
    popperOptions: {
      modifiers: [{
        name: 'arrow',
        options: {
          element: null
        }
      }]
    }
  }, /*#__PURE__*/_react.default.createElement("a", {
    href: "#",
    className: active ? 'active' : '',
    onClick: function onClick(e) {
      e.preventDefault();
      togglePopup();
    },
    onKeyDown: function onKeyDown(e) {
      if (e.key === 'Enter') {
        e.preventDefault();
        togglePopup();
      }
    },
    role: "button"
  }, /*#__PURE__*/_react.default.createElement(_TopBar.default, _extends({}, props, {
    icon: icon,
    rightIcon: rightIcon,
    selectable: selectable,
    className: (0, _classnames.default)({
      'bf-topbar-group-active': tippyVisible
    })
  }), name))));
});
TopBarGroup.propTypes = {
  /** CSS class name(s) */
  className: _propTypes.default.string,
  //** @ignore */
  style: _propTypes.default.object,

  /** Element that opens TopBar.Group */
  name: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.string]),

  /** Content that TopBar.Group will show when opened */
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.element.isRequired), _propTypes.default.element]).isRequired,

  /** Font Awesome icon reference (or string if using library) */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  icon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string]),

  /** Icon position value. This will add space between content and icon based on the position */
  rightIcon: _propTypes.default.bool,

  /** Determines if an item can be selected. Note: This only works when TopBar.Item is a child of TopBar.Group */
  selectable: _propTypes.default.bool
};
var _default = TopBarGroup;
exports.default = _default;
