// @intility/bifrost-react/dist/components/Interactive/Dropdown/Dropdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react2 = _interopRequireDefault(require("@tippyjs/react"));

var _tippy = require("tippy.js");

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Component to display a dropdown on click
 * @visibleName Dropdown
 */
var Dropdown = function Dropdown(_ref) {
  var children = _ref.children,
      className = _ref.className,
      placement = _ref.placement,
      variant = _ref.variant,
      noPadding = _ref.noPadding,
      props = _objectWithoutProperties(_ref, ["children", "className", "placement", "variant", "noPadding"]);

  return /*#__PURE__*/_react.default.createElement(_react2.default, _extends({
    onShow: function onShow() {
      return (0, _tippy.hideAll)();
    },
    maxWidth: "none" // default is 350 for some reason
    ,
    interactive: true,
    className: (0, _classnames.default)(className, 'bf-dropdown', {
      'bf-dropdown-border': variant === 'border',
      'bf-dropdown-nopadding': noPadding
    }),
    placement: placement
  }, props.visible === undefined && {
    trigger: 'click'
  }, props), children);
};

Dropdown.propTypes = {
  /** Element triggering the dropdown */
  children: _propTypes.default.element,

  /** CSS class name(s) */
  className: _propTypes.default.string,

  /** Variant of the dropdown: 'shadow' (default) or 'border' */
  variant: _propTypes.default.oneOf(['shadow', 'border']),

  /** Contents of the Dropdown */
  content: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.string, _propTypes.default.number]),

  /** Placement of the Dropdown (default 'bottom-start') */
  placement: _propTypes.default.oneOf(['auto', 'auto-start', 'auto-end', 'top', 'bottom', 'right', 'left', 'top-start', 'top-end', 'bottom-start', 'bottom-end', 'right-start', 'right-end', 'left-start', 'left-end']),

  /** State for controlled mode */
  visible: _propTypes.default.bool,

  /** Bool to remove padding from content */
  noPadding: _propTypes.default.bool
};
Dropdown.defaultProps = {
  placement: 'bottom-start',
  variant: 'shadow'
};
Dropdown.displayName = 'Dropdown';
var _default = Dropdown;
exports.default = _default;
