// @intility/bifrost-react/dist/components/Navigation/Sidebar/IntilityLogo.internal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Logo;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var itLogo = /*#__PURE__*/_react.default.createElement("svg", {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 18 25",
  "aria-labelledby": "collapsedIntilityLogoSvgTitle",
  role: "img",
  className: "bf-small-it-logo"
}, /*#__PURE__*/_react.default.createElement("title", {
  id: "collapsedIntilityLogoSvgTitle"
}, "Intility logo"), /*#__PURE__*/_react.default.createElement("path", {
  d: "M2.13,1.34A1.35,1.35,0,1,1,3.48,2.69,1.34,1.34,0,0,1,2.13,1.34Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M10.09,8.25H8.32V6.12h1.77V3.73h2.72V6.12H17V8.25h-4.2v4.66a4.43,4.43,0,0,0,.65,2.75,2.52,2.52,0,0,0,2.09.82H17V19h-2.6a4,4,0,0,1-3.07-1.36,4.81,4.81,0,0,1-1.25-3.41Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M2.1,19V8.41H0V6.12H4.82V19Z"
}));

var intilityLogo = /*#__PURE__*/_react.default.createElement("svg", {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 87 25",
  "aria-labelledby": "intilityLogoSvgTitle",
  role: "img",
  className: "bf-full-intility-logo"
}, /*#__PURE__*/_react.default.createElement("title", {
  id: "intilityLogoSvgTitle"
}, "Intility logo"), /*#__PURE__*/_react.default.createElement("path", {
  d: "M48.5,8.26v4.66a4.43,4.43,0,0,0,.65,2.75,2.52,2.52,0,0,0,2.09.82h1.47V19H50.1A4,4,0,0,1,47,17.63a4.85,4.85,0,0,1-1.25-3.42V3.74H48.5V8.26Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M2.14,1.34A1.35,1.35,0,1,1,3.48,2.69,1.34,1.34,0,0,1,2.14,1.34Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M20.07,1.34a1.35,1.35,0,1,1,1.35,1.35A1.34,1.34,0,0,1,20.07,1.34Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M38.1,1.34a1.35,1.35,0,0,1,2.69,0,1.35,1.35,0,1,1-2.69,0Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M56.06,1.34a1.35,1.35,0,1,1,1.35,1.35A1.34,1.34,0,0,1,56.06,1.34Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M73.66,1.34A1.35,1.35,0,1,1,75,2.69,1.36,1.36,0,0,1,73.66,1.34Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M2.11,19V8.41H0V6.12H4.82V19Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M18.54,19V12.74a6,6,0,0,0-.74-3.43,2.9,2.9,0,0,0-2.47-1,4.09,4.09,0,0,0-1.94.56,4.29,4.29,0,0,0-1.56,1.37V19H9.11V6.12h2.77L11.83,8.3a4.72,4.72,0,0,1,4.39-2.25q5,0,5,6.11V19Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M26.09,8.25H24.32V6.12h1.77V3.73h2.72V6.12H33V8.25H28.81v4.66a4.43,4.43,0,0,0,.65,2.75,2.52,2.52,0,0,0,2.09.82H33V19H30.41a4,4,0,0,1-3.07-1.36,4.81,4.81,0,0,1-1.25-3.41Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M64,8.25H62.25V6.12H64V3.73h2.72V6.12h4.2V8.25h-4.2v4.66a4.43,4.43,0,0,0,.65,2.75,2.52,2.52,0,0,0,2.09.82h1.46V19h-2.6a4,4,0,0,1-3.07-1.36A4.81,4.81,0,0,1,64,14.2Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M38.07,19V8.41H36V6.12h4.82V19Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M56,19V8.41h-2.1V6.12h4.82V19Z"
}), /*#__PURE__*/_react.default.createElement("path", {
  d: "M80.52,21.6A4.73,4.73,0,0,1,78.3,24a7.38,7.38,0,0,1-3.77,1V22.56a5.19,5.19,0,0,0,2.91-.79,2.29,2.29,0,0,0,1.21-1.94,7.33,7.33,0,0,0-.44-2.48c-.3-.82-.68-1.82-1.14-3L73.84,6.11h2.77l3.57,9.4,3.18-9.4h2.77Z"
})); // always render both logos, display is toggled in css with .bf-nav-side-collapsed


function Logo() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, intilityLogo, itLogo);
}
