// @intility/bifrost-react/dist/components/Navigation/Sidebar/Sidebar.Item.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _freeRegularSvgIcons = require("@fortawesome/free-regular-svg-icons");

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _Icon = _interopRequireDefault(require("../../Visual/Icon"));

var _useSidebarCollapsed = _interopRequireDefault(require("./hooks/useSidebarCollapsed"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Clickable items for Sidebar
 * @visibleName Sidebar.Item
 */
var SidebarItem = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var children = _ref.children,
      icon = _ref.icon,
      rightIcon = _ref.rightIcon,
      className = _ref.className,
      props = _objectWithoutProperties(_ref, ["children", "icon", "rightIcon", "className"]);

  var collapsed = (0, _useSidebarCollapsed.default)();
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    // todo: stringify react node content
    title: collapsed ? children === null || children === void 0 ? void 0 : children.toString() : undefined
  }, props, {
    className: (0, _classnames.default)(className, 'bf-sidebar-item', {
      'bf-sidebar-item-with-icon': icon,
      'bf-sidebar-item-with-right-icon': rightIcon
    }),
    ref: ref
  }), icon && /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: icon,
    fixedWidth: true,
    className: "bf-sidebar-item-icon"
  }), children, rightIcon ? /*#__PURE__*/_react.default.createElement("span", {
    className: "bf-sidebar-item-righticon"
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: rightIcon,
    fixedWidth: true
  })) :
  /*#__PURE__*/
  // include sublink circle icon by default. hidden with css when not inside a group
  _react.default.createElement("span", {
    className: "bf-sidebar-item-righticon bf-sidebar-group-item-icon"
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _freeRegularSvgIcons.faCircle,
    fixedWidth: true,
    className: "bf-sidebar-group-item-icon-circle"
  }), /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: _freeSolidSvgIcons.faDotCircle,
    fixedWidth: true,
    className: "bf-sidebar-group-item-icon-dot-circle"
  })));
});
SidebarItem.propTypes = {
  /** Sidebar item content (if not a simple string, also supply text version as title) */
  children: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.arrayOf(_propTypes.default.node)]).isRequired,

  /** CSS class name(s) */
  className: _propTypes.default.string,

  /** Font Awesome icon reference (or string if using library) */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  icon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string]),

  /** Font Awesome icon reference (or string if using library) */
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  //@ts-ignore
  rightIcon: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.array, _propTypes.default.string])
};
SidebarItem.displayName = 'Sidebar.Item';
var _default = SidebarItem;
exports.default = _default;
