// @intility/bifrost-react/dist/assets/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "faAngleDoubleLeft", {
  enumerable: true,
  get: function get() {
    return _faAngleDoubleLeft.faAngleDoubleLeft;
  }
});
Object.defineProperty(exports, "faAngleDoubleRight", {
  enumerable: true,
  get: function get() {
    return _faAngleDoubleRight.faAngleDoubleRight;
  }
});
Object.defineProperty(exports, "faAngleDown", {
  enumerable: true,
  get: function get() {
    return _faAngleDown.faAngleDown;
  }
});
Object.defineProperty(exports, "faAngleLeft", {
  enumerable: true,
  get: function get() {
    return _faAngleLeft.faAngleLeft;
  }
});
Object.defineProperty(exports, "faAngleRight", {
  enumerable: true,
  get: function get() {
    return _faAngleRight.faAngleRight;
  }
});
Object.defineProperty(exports, "faAngleUp", {
  enumerable: true,
  get: function get() {
    return _faAngleUp.faAngleUp;
  }
});
Object.defineProperty(exports, "faCheck", {
  enumerable: true,
  get: function get() {
    return _faCheck.faCheck;
  }
});
Object.defineProperty(exports, "faCheckCircle", {
  enumerable: true,
  get: function get() {
    return _faCheckCircle.faCheckCircle;
  }
});
Object.defineProperty(exports, "faExclamationCircle", {
  enumerable: true,
  get: function get() {
    return _faExclamationCircle.faExclamationCircle;
  }
});
Object.defineProperty(exports, "faExclamationTriangle", {
  enumerable: true,
  get: function get() {
    return _faExclamationTriangle.faExclamationTriangle;
  }
});
Object.defineProperty(exports, "faInfoCircle", {
  enumerable: true,
  get: function get() {
    return _faInfoCircle.faInfoCircle;
  }
});
Object.defineProperty(exports, "faSpinnerThird", {
  enumerable: true,
  get: function get() {
    return _faSpinnerThird.faSpinnerThird;
  }
});
Object.defineProperty(exports, "fadSpinnerThird", {
  enumerable: true,
  get: function get() {
    return _duotone.faSpinnerThird;
  }
});
Object.defineProperty(exports, "faTimes", {
  enumerable: true,
  get: function get() {
    return _faTimes.faTimes;
  }
});
Object.defineProperty(exports, "falTimes", {
  enumerable: true,
  get: function get() {
    return _falTimes.falTimes;
  }
});

var _faAngleDoubleLeft = require("./faAngleDoubleLeft");

var _faAngleDoubleRight = require("./faAngleDoubleRight");

var _faAngleDown = require("./faAngleDown");

var _faAngleLeft = require("./faAngleLeft");

var _faAngleRight = require("./faAngleRight");

var _faAngleUp = require("./faAngleUp");

var _faCheck = require("./faCheck");

var _faCheckCircle = require("./faCheckCircle");

var _faExclamationCircle = require("./faExclamationCircle");

var _faExclamationTriangle = require("./faExclamationTriangle");

var _faInfoCircle = require("./faInfoCircle");

var _faSpinnerThird = require("./faSpinnerThird");

var _duotone = require("./duotone");

var _faTimes = require("./faTimes");

var _falTimes = require("./falTimes");
